### PR TITLE
WIP: Proposal to externalize volume plugin idempotency

### PR DESCRIPTION
**Pull request by pmorie**
WIP: Second stab at articulating a proposal for #17281

### CODE REVIEW

**Comment by pmorie** [comment]
@thockin @swagiaal PTAL cc @kubernetes/rh-storage @kubernetes/rh-cluster-infra

File: docs/proposals/volume-idempotency.md [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+The hostPath volume does not implement an idempotency check because its `SetUp` method is a no-op.
+It should be possible to represent that a volume does not have any idempotency concerns.
+
+### Refreshable volume plugins
+
+Some volume plugins are designed to refresh, like the downward API volume plugin.  In order to
+decouple the implementation of refresh and SetUp logic, there should be a Refresh() method that
+plugins can implement and proper machinery to ensure that only plugins that want this need to
+implement it.
```
Comment (swagiaal) [inline docs/proposals/volume-idempotency.md]: This requirement seems orthogonal to this proposal. Also, is there a user-requirement which motivates decoupling refresh from SetUp ? If so I think this should be done within the plugin as it seems like a one off case rather than a general interface.

File: docs/proposals/volume-idempotency.md [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+### Volume plugins that wrap one another
+
+Some volume plugins wrap others:
+
+1.  Secrets, downward API, git repo are based on EmptyDir
+2.  PV claim plugin instantiates instances of the underlying volume plugin
```
Comment (swagiaal) [inline docs/proposals/volume-idempotency.md]: I think PV's can just inherit impotence of the underlying volume. Is there an exception to that ?

File: docs/proposals/volume-idempotency.md [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+Ownership management of volumes must be idempotent for non-refreshable volumes.  It should run once
+and only once after `Setup` has run correctly.  If a volume that uses the 'check for a mountpoint'
+type of idempotency check, and the volume must be remounted, the ownership management should be
+reapplied.
+
+### Ownership management and refreshable volumes
+
+For volume plugins that are refreshable, ownership management must be reapplied if the refresh logic
+makes changes to the state of the volume.  For example, if data changes in the downward API volume,
+the ownership management must be re-applied to that volume.
```
Comment (swagiaal) [inline docs/proposals/volume-idempotency.md]: Hmm.. what is the use case here ? My intuition is that ownership management should be applied to all volumes only once.

File: docs/proposals/volume-idempotency.md [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+	Refreshable          bool
+}
+
+type IdempotencyStrategyType string
+
+const (
+    IdempotencyStrategyNone       IdempotencyStrategyType = "none"
+    IdempotencyStrategyMountpoint IdempotencyStrategyType = "mountPoint"
+    IdempotencyStrategyMarkerFile IdempotencyStrategyType = "markerFile"
+)
```
Comment (swagiaal) [inline docs/proposals/volume-idempotency.md]: hmm.. this seems too expressive. Don't we just need decide whether a volume needs owndership management applied or not ? ie a boolean should suffice.

File: docs/proposals/volume-idempotency.md [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+
+            didRefresh, err := refreshable.Refresh()
+            if err != nil {
+                return err
+            }
+
+            if didRefresh {
+                tryOwnershipManagement = true
+            }
+        }
```
Comment (swagiaal) [inline docs/proposals/volume-idempotency.md]: what if `SetUp` returns a boolean in addition to the error. True if it is a new volume/false if it is an old volume. Plugins can decide what to return in that boolean. Host path always returns false; volumes with a mount dir use the existence of the dir to decide; NFS GlusterFS return always false; empty_dir and ilk return the value of `IsReady()`; block device volumes return a value based on whether they were formatted druing `SetUp()`

File: docs/proposals/volume-idempotency.md (hunk 4 above)
Comment (pmorie) [inline docs/proposals/volume-idempotency.md]: We have to capture the different types of management in order to maintain parity with what the plugins do today.  We have to externalize idempotency checks in order to correctly know when we can attempt ownership management.

**Comment by pmorie** [comment]
Updated to spell out use-case where plugins are backed by another plugin.

File: docs/proposals/volume-idempotency.md (hunk 3 above)
Comment (pmorie) [inline docs/proposals/volume-idempotency.md]: The use-case is volume plugins that are generated by the system that may be refreshed.  For example, the downward API volume plugin may rewrite the data in the volume, in which case, the ownership management should be reapplied.

File: docs/proposals/volume-idempotency.md (hunk 2 above)
Comment (pmorie) [inline docs/proposals/volume-idempotency.md]: Actually, you're right, the PVs use the builders of the underlying plugin, I'll update.

File: docs/proposals/volume-idempotency.md (hunk 1 above)
Comment (pmorie) [inline docs/proposals/volume-idempotency.md]: I'm thinking about it.

Comment (pmorie) [inline docs/proposals/volume-idempotency.md]: I don't think it's exactly orthogonal, but it's arguable that it's not necessary to accomplish the point of this PR.  I'm willing to leave this out for now as this has gotten quite complex.  Going to scan the rest of your feedback, but I'm planning on removing it now.

File: docs/proposals/volume-idempotency.md (hunk 2 above)
Comment (pmorie) [inline docs/proposals/volume-idempotency.md]: That doesn't eliminate the need to handle this for secrets, git repo, etc.

File: docs/proposals/volume-idempotency.md [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+
+1.  It must be possible to express that a volume plugin has no idempotency check
+2.  It must be possible to express that a volume plugin uses either a mountpoint check or marker
+    file
+3.  Marker file volumes must have the marker file created after `SetUp` runs successfully
+4.  Refresh logic for refreshable plugins must be factored out of `SetUp`, and be run only if
+    `SetUp` has run successfully.
+5.  Ownership management must run once for each time `Setup` is invoked or refresh logic changes
+    the state of a volume
+6.  Volume plugins must be able to trigger the externalized Setup flow to support correct composition
```
Comment (ncdc) [inline docs/proposals/volume-idempotency.md]: Put Setup in backticks?

File: docs/proposals/volume-idempotency.md [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+    of plugins
+
+The volume `Attributes` structure is the appropriate place to store information the properties we
```
Comment (ncdc) [inline docs/proposals/volume-idempotency.md]: information *for*?

File: docs/proposals/volume-idempotency.md (hunk 3 above)
Comment (swagiaal) [inline docs/proposals/volume-idempotency.md]: I see. I think you had mentioned that you would remove the setgid from ownership management. Is that what creates the need for the above ?

I have a suggestion: what if the downward API plugin creates new files using the root ownership of the volume ? That would:
- Eliminate the need for `Refresh()` simplifying this proposal.
- Eliminate the window of incorrect ownership between the creation of the file and the next `Refresh()` cycle

File: docs/proposals/volume-idempotency.md (hunk 2 above)
Comment (swagiaal) [inline docs/proposals/volume-idempotency.md]: Thats right

**Comment by pmorie** [comment]
Closing this out; going to open a new PR with the reverse of this proposal (internalize)
